### PR TITLE
Cluster

### PR DESCRIPTION
**Pull request by ettenhup**
Hi Benjamin, 
I cleaned up some parts and implemented some more options, e.g. an indicator function for the cluster-similarity, the option to symmetrize the transition matrix with "--mcl-symmetrize". I have moved the options "--mcl-restart" to "--cluster-restart" and the "--mcl-graph-file" to "--cluster-graph-file" as these should be able to work with any clustering method if the ClusterOutput option is chosen together with the sparse_matrix_stream object.

Cheers,
Patrick

### CODE REVIEW

File: src/cluster/mcl.cpp [hunk 1]
```diff
@@ -265,42 +250,69 @@ void MCL::markov_process(Eigen::MatrixXf* m, float inflation, float expansion, u
...
+		uint32_t upper_limit =  min(my_counter+my_chunk_size, nComponentsLt1);
+		vector<vector<uint32_t>*> loc_i;
+		for(uint32_t chunk_counter = my_counter; chunk_counter<upper_limit; chunk_counter++){
+			loc_i.push_back(&indices[sort_order[chunk_counter]]);
+		}
+		vector<vector<Eigen::Triplet<float>>> loc_c = ms->collect_components(&loc_i, 0);
+		uint32_t ichunk = 0;
+		generate(sparsities.begin()+my_counter, sparsities.begin()+upper_limit, [my_counter, ichunk, &sort_order, &indices, &loc_c]() mutable { 
+					const uint32_t iComponent = sort_order[my_counter++];
+					printf("ichunk %u\n",ichunk);
```
Comment (bbuchfink) [inline src/cluster/mcl.cpp:284]: Please write to stderr since stdout is reserved for output.

File: src/cluster/sparse_matrix_stream.h [hunk 2]
```diff
@@ -106,6 +108,12 @@ class SparseMatrixStream : public Consumer {
 				const uint32_t subject = *(uint32_t*) ptr;
 				ptr += sizeof(uint32_t);
 				const double value = *(double*) ptr;
+				if(!warned && (value > std::numeric_limits<T>::max() || value < std::numeric_limits<T>::min())){
+					printf("\n");
```
Comment (bbuchfink) [inline src/cluster/sparse_matrix_stream.h:112]: Please write to stderr since stdout is reserved for output.

Comment (ettenhup) [reply src/cluster/sparse_matrix_stream.h:112]: Will do, thanks.

File: src/cluster/mcl.cpp (hunk 1 above)
Comment (ettenhup) [reply src/cluster/mcl.cpp:284]: This should not have been there at all, I will remove it. Good catch.

Comment (ettenhup) [reply src/cluster/mcl.cpp:284]: Hm, I am pretty sure I had removed this prior to the commit. I am wondering whether there is a bug in github's diff or whether the removal is actually missing. We need to double-check!

**Comment by ettenhup** [comment]
Should be fixed now.

**Pull request merged**